### ISSUE TITLE
[IDEA] Use a boxed std::any::Any for PyObjectPayload instead of an enum

### ISSUE THREAD

**Issue by coolreader18** [RustPython/RustPython#535]
[Documentation on the `Any` trait](https://doc.rust-lang.org/stable/std/any/trait.Any.html)

This came about from thinking about having a `objpromise` class in WASM. How would an instance of that be stored in the VM? If it were some type we were integrating into the core `rustpython_vm`, we'd just add another variant to `PyObjectPayload`, but that couldn't be done for a WASM type. For that matter, what about any other library trying to integrate a Python representation of their Rust struct into RustPython from the outside? Therefore, I think it might be an option to change `PyObject` to have `payload` be of type `Box<dyn any::Any>`, and `get_value()` fns or similar could do something like this to check:

```rust
struct ObjFloat {
    value: f64,
}
fn get_value(obj: &PyObjectRef) -> f64 {
    if let Some(float) = obj.borrow().payload.downcast_ref::<ObjFloat>() {
        float.value
    } else {
        panic!("Inner error getting float");
    }
}
```

This might even help with memory consumption; instead of storing 336 bytes per payload, it would store 16, not counting runtime costs of `Box` dereferencing and `Any` checking.

**Comment by OddCoincidence** [comment]
Very interesting idea! If we revisit object representation, we should also take into account some of the suggestions in https://github.com/RustPython/RustPython/issues/371.

**Comment by coolreader18** [comment]
This has been resolved with #653.
